deltacat/compute/compactor/utils/repartition_session.py

```diff
@@ -0,0 +1,168 @@
+        **list_deltas_kwargs,
+    )
+
+    uniform_deltas = []
+    for delta in deltas:
+        # limit the input deltas to fit on this cluster and convert them to
+        # annotated deltas of equivalent size for easy parallel distribution
+        uniform_deltas_part = DeltaAnnotated.rebatch(
+            [DeltaAnnotated.of(delta)],
+            200 * 2**20,  # 200 MiB compressed data per task,
+            min_file_counts=1000,
+        )
+        uniform_deltas.extend(uniform_deltas_part)
+
+    logger.info(f"Retrieved a total of {len(uniform_deltas)} uniform deltas.")
+
+    max_parallelism = cluster_cpus
+    # create a new stream for this round
+    compacted_stream_locator = destination_partition_locator.stream_locator
+    stream = deltacat_storage.get_stream(
+        compacted_stream_locator.namespace,
+        compacted_stream_locator.table_name,
+        compacted_stream_locator.table_version,
+    )
+    partition = deltacat_storage.stage_partition(
+        stream,
+        destination_partition_locator.partition_values,
+    )
+    new_compacted_partition_locator = partition.locator
+    repar_start = time.time()
+    repar_tasks_pending = invoke_parallel(
+        items=uniform_deltas,
+        ray_task=repar.repartition,
+        max_parallelism=max_parallelism,
+        options_provider=round_robin_opt_provider,
+        repartition_type=repartition_type,
+        repartition_args=repartition_args,
+        destination_partition=partition,
+        enable_profiler=False,
+        metrics_config=None,
+        read_kwargs_provider=read_kwargs_provider,
+        deltacat_storage=deltacat_storage,
+    )
+    logger.info(f"Getting {len(repar_tasks_pending)} task results...")
+    repar_results: List[RePartitionResult] = ray.get(repar_tasks_pending)
+    repar_results: List[Delta] = [rp.range_deltas for rp in repar_results]
+    transposed = list(itertools.zip_longest(*repar_results, fillvalue=None))
+    ordered_deltas: List[Delta] = [
+        i for sublist in transposed for i in sublist if i is not None
+    ]
+    repar_end = time.time()
+    logger.info(f"repartition {repar_end - repar_start} seconds")
+
+    logger.info(f"Got {len(ordered_deltas)} task results.")
+    # ordered_deltas are ordered as [cold1, cold2, coldN, hot1, hot2, hotN]
+    merged_delta = Delta.merge_deltas(ordered_deltas)
+    compacted_delta = deltacat_storage.commit_delta(
+        merged_delta, properties=kwargs.get("properties", {})
+    )
+    deltacat_storage.commit_partition(partition)
+    logger.info(f"Committed final delta: {compacted_delta}")
+    logger.info(f"Job run completed successfully!")
+    new_compacted_delta_locator = DeltaLocator.of(
+        new_compacted_partition_locator,
+        compacted_delta.stream_position,
+    )
+    bit_width_of_sort_keys = SortKey.validate_sort_keys(
+        source_partition_locator,
+        sort_keys,
+        deltacat_storage,
+    )
+    new_round_completion_info = RoundCompletionInfo.of(
```

Comment:
Do we need to write round completion info for repartitioning? Will it get confused to be actual compaction is done on this partition and stream position?

Reply:
also, repartition is similar to a rebase compaction assuming no new deltas

Reply:
I am not sure, that idea does not resonate very well with what repartition does. It will not `compact` at all, so it cannot be thought of as a compaction.

Reply:
It does 'compact', right? it compact cold data together and hot data together; In rebase, it compact small files into larger files, by setting records per files;

And with further optimization in next release, repartition can be a better version of rebase, as we can figure out how to compact cold data together into less number of cold files.

Reply:
The Iceberg domain word is [copy-on-write](https://aws.amazon.com/blogs/big-data/use-apache-iceberg-in-a-data-lake-to-support-incremental-data-processing/#Copy-on-write%20vs.%20read-on-merge), where we duplicate the files that were updated upon writes. One critical requirement is that it works on new logs while re-partitioning works on current state of data in the table though I somewhat agree goals are same.